FOFR-35_Implementation_of_Keycloak_Token_Handling_in_Offline_Mode

sandeep-aot opened the pull request:
**NOTE:- This PR has been created for review purposes only by the forms flow product team for inputs regarding this approach and not for any merging to any branch.**

Includes:- Handling token refresh while offine to avoid any page refresh which will display the "No Internet page" to the customer.

sandeep-aot committed:
FOFR-35_Implementation_of_Keycloak_Token_Handling_in_Offline_Mode

NOTE:- This PR has been created for review purposes only by the forms flow product team for inputs regarding this approach and not for any merging to any branch.

Includes:- Handling token refresh while offine to avoid any page refresh which will display the "No Internet page" to the customer.

arun-s-aot reviewed (approved):
Looks ok to me

arun-s-aot commented on forms-flow-service/src/keycloak/KeycloakService.ts:
```diff
@@ -60,11 +60,79 @@ import Keycloak, {
+          return;
+        }
+    
+        this.kc
+          ?.updateToken(-1)
```
@sandeep-aot What will happen if we give -1 ? if we are planning to run it once, i think we should be passing 0

sandeep-aot replied:
@arun-s-aot This part of the code `this.kc
          ?.updateToken(-1)` is what i have retained from our existing working code.In the existing working code which is meant for online,it seems it was used for token refresh regardless of the current expiration time.
This code will deal with the offline check
`if (!navigator.onLine) {
      console.debug("Offline: Skipping token refresh.");
      return;
    }
`
I have also replaced `this.logout();` with `this.handleTokenRefreshFailure();` which will prevent the user logout should the update fail and error out.

This piece of code was added to handle the token refresh as soon as the event listener picks up the online event
` private retryTokenRefresh = (): void => {
      console.log("Back online: Retrying token refresh.");
      //Reset flag after retry
      this.isWaitingForOnline = false;
      this.refreshToken();
    };`
Could you please also review these parts of the code that i have highlighted so that we can ensure that these changes would not affect the existing online functionality adversly?

sandeep-aot committed:
FOFR_35_Keycloak_fix_for_offline_page_refresh

This checkin removes some redundant comments on the new code and also the commented earlier code.

sandeep-aot committed:
FOFR-35_keycloak_offline_issue_fix

Removes a comment inside the code
